server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardWriteLoadDistributionMetrics.java

```diff
@@ -0,0 +1,166 @@
+        final var shardWriteLoads = clusterInfo.getShardWriteLoads();
+        final var ingestNodeCount = (int) clusterState.nodes().stream().filter(DiscoveryNode::isIngestNode).count();
+        final var writeLoadDistributionValues = new ArrayList<DoubleWithAttributes>(percentiles.length * ingestNodeCount);
+        final var writeLoadPrioritisationThresholdValues = new ArrayList<DoubleWithAttributes>(ingestNodeCount);
+        final var shardCountsExceedingPrioritisationThresholdValues = new ArrayList<LongWithAttributes>(ingestNodeCount);
+        for (RoutingNode routingNode : clusterState.getRoutingNodes()) {
+            final var node = routingNode.node();
+            if (node == null || node.isIngestNode() == false) {
+                continue;
+            }
+
+            double maxShardWriteLoad = Double.NEGATIVE_INFINITY;
+
+            shardWeightHistogram.reset();
+            try {
+                for (ShardRouting shardRouting : routingNode) {
+                    Double writeLoad = shardWriteLoads.get(shardRouting.shardId());
+                    if (writeLoad != null) {
+                        shardWeightHistogram.recordValue(writeLoad);
+                        maxShardWriteLoad = Math.max(maxShardWriteLoad, writeLoad);
+                    }
+                }
+            } catch (ArrayIndexOutOfBoundsException e) {
+                // This shouldn't happen because our histogram should be auto-resizing, but just in case
+                logger.error("Failed to record shard write load distribution metrics for node " + node.getName(), e);
+                continue;
+            }
+
+            for (double percentile : percentiles) {
+                writeLoadDistributionValues.add(
+                    new DoubleWithAttributes(
+                        shardWeightHistogram.getValueAtPercentile(percentile),
+                        getAttributesForPercentile(node, percentile)
+                    )
+                );
+            }
+            final var nodeAttrs = getAttributesForNode(node);
+            final double prioritisationThreshold = BalancedShardsAllocator.Balancer.PrioritiseByShardWriteLoadComparator.THRESHOLD_RATIO
+                * maxShardWriteLoad;
+            writeLoadPrioritisationThresholdValues.add(new DoubleWithAttributes(prioritisationThreshold, nodeAttrs));
+
+            final long shardsExceedingThreshold = (long) shardWeightHistogram.getCountBetweenValues(
+                prioritisationThreshold,
+                Double.MAX_VALUE
+            );
+            shardCountsExceedingPrioritisationThresholdValues.add(new LongWithAttributes(shardsExceedingThreshold, nodeAttrs));
+        }
+
+        lastWriteLoadDistributionValues.set(writeLoadDistributionValues);
+        lastWriteLoadPrioritisationThresholdValues.set(writeLoadPrioritisationThresholdValues);
+        lastShardCountExceedingPrioritisationThresholdValues.set(shardCountsExceedingPrioritisationThresholdValues);
```

Comment:
We will calculate these values twice between each collection, that seems excessive/wasteful. I wonder if its worth optimising so it happens once per collection.